### DISCUSSION TITLE
Friend class

### DISCUSSION THREAD

**Discussion by alex4998** [dotnet/csharplang#5734] [Language Ideas] [answered]
For decades I've been missing friend class declaration in C#, similar to friend in C++. This topic was discussed hundreds of times and there are many people against it, but it would really help to make the isolation better not worse. internal class is not the same. nested class is not the same. I need a way to make one particular class being able to access other particular class protected/private members. 

In the absence of this ability, I'm forced to make certain members public or internal, it doesn't matter, only to call them from one particular place in another class. I would like to hide those members for everybody else so that nobody, working on this assembly or on assembly using this one, would think that the members are exposed to be called in particular scenarios. Unfortunately, they were exposed because there is no way to proper hide them.

The use case is quite obvious. Let's say there is a data provider class. It manages two aspects of data handling. One aspect that needs to be public is data fetching, data updating and so on. The other aspect is changing/updating data schema. Functionally it is part of the data provider which knows about the data, knows how the data is stored and everything else in this regard. Structurally, the schema managing aspect is not supposed to be exposed to everybody, it is supposed to be exposed only to a schema controller class, which exposes high level interface and calling low level data schema managing methods on the data provider.

We can turn this requirement at different angles but anyway, there will be a need for low level functional members which are supposed to be exposed only to a particular controller or manager but at the same time there will be other members which need to be exposed to the wide audience.

**Comment by bernd5** [comment]
You could define an interface for each individual aspect (in this case two) and implement all functions explicitly.
This would allow access to the necessary members where needed while keeping them not public...

**Comment by HaloFour** [comment, marked as answer]
There are existing discussions regarding friend classes: #2073
